File: src/test/scala/com/github/kayjamlang/tests/expressions/operation/InFunctionContainerOperationTest.scala
package com.github.kayjamlang.tests.expressions.operation

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{CallOrCreateExpression, OperationExpression, ValueExpression}
import com.github.kayjamlang.core.expressions.data.Operation
import org.junit.Assert._
import org.junit.{BeforeClass, Test}


object InFunctionContainerOperationTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer("equals(2+2)"))
  }
}

class InFunctionContainerOperationTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = InFunctionContainerOperationTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[CallOrCreateExpression], expression.getClass)
    val callOrCreateExpression = expression.asInstanceOf[CallOrCreateExpression]
    assertEquals(1, callOrCreateExpression.arguments.size)
    assertSame(classOf[OperationExpression], callOrCreateExpression.arguments.head.getClass)
    val operationExpression = callOrCreateExpression.arguments.head.asInstanceOf[OperationExpression]
    assertEquals(Operation.PLUS, operationExpression.operation)
    assertSame(classOf[ValueExpression], operationExpression.left.getClass)
    assertSame(classOf[ValueExpression], operationExpression.right.getClass)
  }
}

File: src/test/java/com/github/kayjamlang/tests/expressions/operation/InFunctionContainerOperationTest.java
package com.github.kayjamlang.tests.expressions.operation;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.data.Operation;
import com.github.kayjamlang.core.expressions.CallOrCreateExpression;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.OperationExpression;
import com.github.kayjamlang.tests.TestsUtils;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class InFunctionContainerOperationTest {

    @Test
    public void test() throws Exception {
        Expression expression = TestsUtils.parse("equals(2+2)");

        assertNotNull(expression);
        assertSame(CallOrCreateExpression.class, expression.getClass());

        CallOrCreateExpression callOrCreateExpression = (CallOrCreateExpression) expression;
        assertEquals(1, callOrCreateExpression.arguments.size());
        assertSame(OperationExpression.class, callOrCreateExpression.arguments.get(0).getClass());

        OperationExpression operationExpression = (OperationExpression) callOrCreateExpression.arguments.get(0);
        assertEquals(Operation.PLUS, operationExpression.operation);
        assertSame(ValueExpression.class, operationExpression.left.getClass());
        assertSame(ValueExpression.class, operationExpression.right.getClass());
    }
}
